### PR TITLE
Resolve ILLink warnings on System.Private.DataContractSerialization Part 1; Json)

### PR DESCRIPTION
**Pull request by joperezr**
This is part 1 of resolving the warnings for System.Private.DataContractSerialization. It is covering all of the warnings from the Json side, the next step will be to cover the Xml side.

Given the space is serialization, these changes mostly are about adding RequiresUnreferencedCode to all the stack that uses reflection to serialize/deserialize objects comming from the input.

cc: @eerhardt @krwq @vitek-karas

### CODE REVIEW

**Comment by joperezr** [comment]
As said above, the initial approach was to just mark everything as RequiredUnreferencedCode given the fact that this is a serializer, but due to the nature of DataContractSerialization and the fact that you have to always create the serializer passing in the expected types in the object graph, I'm checking to see if we can actually annotate this accordingly in order to make this serializer safe. I'll try a few things and update the PR if necessary

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs [hunk 1]
```diff
@@ -497,21 +511,27 @@ internal sealed class DataContractJsonSerializerImpl : XmlObjectSerializer
         private DateTimeFormat? _dateTimeFormat;
         private bool _useSimpleDictionaryFormat;
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
```
Comment (joperezr) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:514]: After further inspection, looks like this suppression is wrong. Even though here we are not passing in knownTypes directly to the serializer, this is not taking into account the fact that the type you pass in may have a KnownTypeAttribute to it which would make it go through the unsafe codePath.

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatWriterGenerator.cs [hunk 2]
```diff
@@ -98,9 +103,11 @@ internal JsonFormatCollectionWriterDelegate GenerateCollectionWriter(CollectionD
                 return (JsonFormatCollectionWriterDelegate)_ilg.EndMethod();
             }
 
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
```
Comment (eerhardt) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatWriterGenerator.cs:106]: Can we add a new helper/utility method named `GetInvokeMethod(Type delegateType)` that suppresses this message and does the `Debug.Assert`? That way we are only suppressing a single GetMethod call, and aren't over suppressing other warnings.

**Review by eerhardt — APPROVED** [review]
Just one suggestion to factor the Delegate GetMethod("Invoke") out into a helper method. Once that is addressed this looks good.

**Review by eerhardt — APPROVED** [review]

**Pull request merged**